### ISSUE TITLE
Initial Setup

### ISSUE THREAD

**Issue by Richygecko** [tuxntoast/eg4-ll#2]
Hi, I've been using a previous version of this code with some tweaks but on separate USB channels. I've installed this new version without too many issues for a single battery but I'm a bit confused with the wiring setup as described for multiple batteries:

-# Battery Tested on:
-# 2x Eg4 LL 12v 400 AH
-# One RS232 Cable to USB is needed when connecting to the Cerbo GX 
-# A networking cable should be connected between each of the BMS RS485 Ports and the other BMS units
-# The master unit or first unit should have a Dip Switch ID set to 16 (12v) or 64 (24v and 48v)
-# All other BMS should have a Dip switch setting of 1 - 16 (12v) or 1 - 63 (24v and 48v)

are you wiring from each BMS into a single RS485<->USB -> Cerbo GX?

For reference I'm using 2x Eg4ll 24v 200ah batteries with 4 position dip switches 
Master DDDD
Slave1 DDDU

Cheers
Nathan

**Comment by tuxntoast** [comment]
I have a battery array of 2. Battery ID of 16 is got a USB to RS485 to connect to the Cerbo, and a cat 5 to battery with ID of 1. 
Battery with ID of 1 is not directly connected to the Cerbo. 

The driver is able to use the cable that is connected to ID 16, to also reach all other BMS that are chained.

There was someone with a 24 or 48 volt system that informed me that the master ID should be 64, and not 16. Not sure if that is just for the new version or it it applied to the prior gen units as well. I have a release that I am working on testing that would allow for commands to be generated with any master ID. The latest version on this repo should include that support. 

If that is the dips on your units, then your master would be 16, and the only slave in the array is ID 1.

**Comment by Richygecko** [comment]
It appears to be working when changing the batteryPackId to either 16 or 15 (I'm using DDDD and UUUU as ID =1 was throwing all kinds of errors) However when using  batteryPackId = [ 16, 15 ] Im getting: 

2024-09-20 18:27:49.552140500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Cell

Many apologies as it's been a lifetime since I've been this deep into code.

Cheers,

Nathan

**Comment by tuxntoast** [comment]
So according to the bms documentation you need the mast to be with an id of 16 (DDDD or all the dip switches down), then any other bms should start counting from 1 (DDDU or just the last dip switche in the bank to up) I did do testing a while ago, and did not seem to find that a master of 16 was required, and I was able to poll any bms id as long as the id was defined in the driver correctly, and the bms is correctly assigned that id on the front of the unit. You do need to power cycle the bms each time you want to change the id The log message you included would be more helpful if you gave more more lines I have been working on trying to address no reply from, and personally seeing about 4-6 per hour… but when you figure that the polling is every 5 seconds, missing 4 bms updates in an hour is not that bad! Generally the next cycle pick up that bms and any data that might have changed since the last time it was fetched. When the driver is starting it really only needs to connect to the master the other are not required for the driver to get a lock. Based on your single log message I can’t tell if this no reply is during first connection , or just over time. Defaults the code is set to only print that single error message, and other wise write nothing else to the log. If you want to see if it is actually getting both batteries you can set statusLogging = True, reload /restart the driver, and then there would be a status printed each polling cycle. That stays will print stats on all batteries that it connected to during that polling cycle Hope that helps , if not can you get me more logs? Also might check your dip switches are correct, and the right numbers. Else try a power cycle on the bms. I have seen where a unit just no longer is responding to polling requests. And only way to fix is with a power cycle Sent from my iPhoneOn Sep 20, 2024, at 11:37 AM, Richygecko ***@***.***> wrote:﻿
It appears to be working when changing the batteryPackId to either 16 or 15 (I'm using DDDD and UUUU as ID =1 was throwing all kinds of errors) However when using  batteryPackId = [ 16, 15 ] Im getting:
2024-09-20 18:27:49.552140500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Cell
Many apologies as it's been a lifetime since I've been this deep into code.
Cheers,
Nathan

**Comment by waconiglio** [comment]
You may have already done this correctly, but my understanding is that only the master connects to the Cerbo from the port labeled RS485 at top right. All other daisy-chained communication happens using the two ports on the lower left of the batteries.

**Comment by Richygecko** [comment]
Attached is a picture of my batteries and a snippet from the manual. In my manual it says to put the Dip switches to ID=1 for master? I know there is a lot of confusion around this. But 16 seems to work fine and all but ID=1 returns data. Using BMS tools i can get the make and model for all ID's 2-16 but not ID=1 it returns some giberish. 

Here's a picture of my batteries. IGNORE THE DIP SETTINGS AND CABLING

![20240918_144433](https://github.com/user-attachments/assets/d2d1bd13-1e81-4c07-9e47-d13df5d311bb)

![Screenshot 2024-09-21 152445](https://github.com/user-attachments/assets/bd50954b-4bc1-4206-ac32-9a9d8eda9741)

![Screenshot 2024-09-21 165536](https://github.com/user-attachments/assets/061fb303-d575-47d9-be7c-5af2f88aa3a4)

![Screenshot 2024-09-21 165812](https://github.com/user-attachments/assets/527e9f32-46cc-4c1e-94f9-ab49fe38e0af)

![Screenshot 2024-09-21 170300](https://github.com/user-attachments/assets/55addb93-fc4f-4ebe-a2a3-62540fa4f138)

![Screenshot 2024-09-21 170324](https://github.com/user-attachments/assets/62cccfaf-35f3-4221-b019-14c9cf633aa8)

### batteryPackId = [ 16, 1 ]

2024-09-21 21:06:12.033898500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:06:12.035944500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:06:12.036607500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705
2024-09-21 21:06:28.052218500 INFO:SerialBattery:-- Testing BMS: 1 of 3 rounds
2024-09-21 21:06:28.053114500 INFO:SerialBattery:Testing Daly at address "\x40"
2024-09-21 21:06:28.185898500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:06:28.186878500 INFO:SerialBattery:Testing Daly at address "\x80"
2024-09-21 21:06:28.318844500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:06:28.319786500 INFO:SerialBattery:Testing Ecs
2024-09-21 21:06:28.378823500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:06:28.380103500 INFO:SerialBattery:Testing EG4_Lifepower
2024-09-21 21:06:28.665230500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:06:28.666259500 INFO:SerialBattery:Testing EG4_LL at address "\x7c"
2024-09-21 21:06:29.695408500 ERROR:SerialBattery:No Reply - BMS ID:1 Command-Hardware
2024-09-21 21:06:29.696675500 INFO:SerialBattery:Connected to BMS ID's: {16: True}
2024-09-21 21:06:29.842896500 INFO:SerialBattery:Connection established to EG4_LL
2024-09-21 21:06:32.953482500 INFO:SerialBattery:Found existing battery with DeviceInstance = 4
2024-09-21 21:06:33.217851500 INFO:SerialBattery:DeviceInstance = 4
2024-09-21 21:06:33.219564500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_4.pid
2024-09-21 21:06:33.222161500 INFO:SerialBattery:Used DeviceInstances = ['4', '3', '2', '1']
2024-09-21 21:06:33.222171500 INFO:SerialBattery:com.victronenergy.battery.ttyUSB0
**2024-09-21 21:06:33.733201500 Traceback (most recent call last):
2024-09-21 21:06:33.733210500   File "/opt/victronenergy/dbus-serialbattery/dbus-serialbattery.py", line 291, in <module>
2024-09-21 21:06:33.733214500     main()
2024-09-21 21:06:33.733216500   File "/opt/victronenergy/dbus-serialbattery/dbus-serialbattery.py", line 240, in main
2024-09-21 21:06:33.733219500     if not helper.setup_vedbus():
2024-09-21 21:06:33.733222500   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 489, in setup_vedbus
2024-09-21 21:06:33.733225500     if not self.battery.get_settings():
2024-09-21 21:06:33.733403500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 135, in get_settings
2024-09-21 21:06:33.733409500     hw_reply = self.read_hw_details(ser, id)
2024-09-21 21:06:33.733411500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 188, in read_hw_details
2024-09-21 21:06:33.733415500     battery.update({"hw_make" : result[2:25].decode("utf-8")})
2024-09-21 21:06:33.733418500 UnicodeDecodeError: 'utf-8' codec can't decode byte 0x84 in position 20: invalid start byte**
2024-09-21 21:06:51.732510500 INFO:SerialBattery:
2024-09-21 21:06:51.733218500 INFO:SerialBattery:Starting dbus-serialbattery


### batteryPackId = [ 16, 1 ]
statuslogger = True

2024-09-21 21:17:22.909864500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:17:22.917742500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:17:22.918786500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705
2024-09-21 21:17:38.935921500 INFO:SerialBattery:-- Testing BMS: 1 of 3 rounds
2024-09-21 21:17:38.937003500 INFO:SerialBattery:Testing Daly at address "\x40"
2024-09-21 21:17:39.071274500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:17:39.072239500 INFO:SerialBattery:Testing Daly at address "\x80"
2024-09-21 21:17:39.204368500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:17:39.205142500 INFO:SerialBattery:Testing Ecs
2024-09-21 21:17:39.263691500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:17:39.264408500 INFO:SerialBattery:Testing EG4_Lifepower
2024-09-21 21:17:39.544580500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:17:39.545702500 INFO:SerialBattery:Testing EG4_LL at address "\x7c"
2024-09-21 21:17:40.612351500 ERROR:SerialBattery:No Reply - BMS ID:1 Command-Hardware
2024-09-21 21:17:40.613676500 INFO:SerialBattery:Connected to BMS ID's: {16: True}
2024-09-21 21:17:40.724155500 INFO:SerialBattery:Connection established to EG4_LL
2024-09-21 21:17:43.794093500 INFO:SerialBattery:Found existing battery with DeviceInstance = 4
2024-09-21 21:17:44.053390500 INFO:SerialBattery:DeviceInstance = 4
2024-09-21 21:17:44.055136500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_4.pid
2024-09-21 21:17:44.055861500 INFO:SerialBattery:Used DeviceInstances = ['4', '3', '2', '1']
2024-09-21 21:17:44.056702500 INFO:SerialBattery:com.victronenergy.battery.ttyUSB0
**2024-09-21 21:17:45.064443500 Traceback (most recent call last):
2024-09-21 21:17:45.064452500   File "/opt/victronenergy/dbus-serialbattery/dbus-serialbattery.py", line 291, in <module>
2024-09-21 21:17:45.064457500     main()
2024-09-21 21:17:45.064458500   File "/opt/victronenergy/dbus-serialbattery/dbus-serialbattery.py", line 240, in main
2024-09-21 21:17:45.064461500     if not helper.setup_vedbus():
2024-09-21 21:17:45.064464500   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 489, in setup_vedbus
2024-09-21 21:17:45.064467500     if not self.battery.get_settings():
2024-09-21 21:17:45.064600500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 136, in get_settings
2024-09-21 21:17:45.064605500     cell_reply = self.read_cell_details(ser, id)
2024-09-21 21:17:45.064608500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 239, in read_cell_details
2024-09-21 21:17:45.064611500     battery.update({"cell_max" : max(cellVoltageList)})
2024-09-21 21:17:45.064614500 ValueError: max() arg is an empty sequence**
2024-09-21 21:18:03.121277500 INFO:SerialBattery:
2024-09-21 21:18:03.122036500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:18:03.124148500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:18:03.124797500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705

batteryPackId = [ 16, 15 ]
statuslogger = True

2024-09-21 21:31:17.110205500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:31:17.112574500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:31:17.113239500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705
2024-09-21 21:31:33.128590500 INFO:SerialBattery:-- Testing BMS: 1 of 3 rounds
2024-09-21 21:31:33.129713500 INFO:SerialBattery:Testing Daly at address "\x40"
2024-09-21 21:31:33.262861500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:31:33.263865500 INFO:SerialBattery:Testing Daly at address "\x80"
2024-09-21 21:31:33.395503500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:31:33.396425500 INFO:SerialBattery:Testing Ecs
2024-09-21 21:31:33.462132500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:31:33.462832500 INFO:SerialBattery:Testing EG4_Lifepower
2024-09-21 21:31:33.747478500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:31:33.748629500 INFO:SerialBattery:Testing EG4_LL at address "\x7c"
2024-09-21 21:31:34.778497500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Hardware
2024-09-21 21:31:34.779679500 INFO:SerialBattery:Connected to BMS ID's: {16: True}
2024-09-21 21:31:34.924980500 INFO:SerialBattery:Connection established to EG4_LL
2024-09-21 21:31:37.959677500 INFO:SerialBattery:Found existing battery with DeviceInstance = 4
2024-09-21 21:31:38.231893500 INFO:SerialBattery:DeviceInstance = 4
2024-09-21 21:31:38.233784500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_4.pid
2024-09-21 21:31:38.234542500 INFO:SerialBattery:Used DeviceInstances = ['4', '3', '2', '1']
2024-09-21 21:31:38.235386500 INFO:SerialBattery:com.victronenergy.battery.ttyUSB0
2024-09-21 21:31:39.694826500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Hardware
2024-09-21 21:31:39.980527500 ERROR:SerialBattery:ERROR >>> Problem with battery set up at /dev/ttyUSB0
2024-09-21 21:31:56.992718500 INFO:SerialBattery:
2024-09-21 21:31:56.993491500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:31:56.995536500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:31:56.996174500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705
2024-09-21 21:32:13.010153500 INFO:SerialBattery:-- Testing BMS: 1 of 3 rounds
2024-09-21 21:32:13.011100500 INFO:SerialBattery:Testing Daly at address "\x40"
2024-09-21 21:32:13.145883500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:32:13.146842500 INFO:SerialBattery:Testing Daly at address "\x80"
2024-09-21 21:32:13.278631500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:32:13.279403500 INFO:SerialBattery:Testing Ecs
2024-09-21 21:32:13.338294500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:32:13.339006500 INFO:SerialBattery:Testing EG4_Lifepower
2024-09-21 21:32:13.609592500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:32:13.610774500 INFO:SerialBattery:Testing EG4_LL at address "\x7c"
2024-09-21 21:32:14.641537500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Hardware
2024-09-21 21:32:14.642800500 INFO:SerialBattery:Connected to BMS ID's: {16: True}
2024-09-21 21:32:14.788082500 INFO:SerialBattery:Connection established to EG4_LL
2024-09-21 21:32:17.812863500 INFO:SerialBattery:Found existing battery with DeviceInstance = 4
2024-09-21 21:32:18.070404500 INFO:SerialBattery:DeviceInstance = 4
2024-09-21 21:32:18.072488500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_4.pid
2024-09-21 21:32:18.073213500 INFO:SerialBattery:Used DeviceInstances = ['4', '3', '2', '1']
2024-09-21 21:32:18.074021500 INFO:SerialBattery:com.victronenergy.battery.ttyUSB0
2024-09-21 21:32:19.078859500 INFO:SerialBattery:===== HW Info =====
2024-09-21 21:32:19.079553500 INFO:SerialBattery:Battery Make/Model: .GRW-08S16P200Ah-100A-2
2024-09-21 21:32:19.080280500 INFO:SerialBattery:Hardware Version: Z15T05
2024-09-21 21:32:19.080980500 INFO:SerialBattery:Serial Number: 082420020020000_16
2024-09-21 21:32:19.081695500 INFO:SerialBattery:Serial Number: 082420020020000_15
2024-09-21 21:32:19.082328500 INFO:SerialBattery:===== Temp =====
2024-09-21 21:32:19.082998500 INFO:SerialBattery:Temp 1: 30c | Temp 2: 27c | Temp Mos: 28c
2024-09-21 21:32:19.083597500 INFO:SerialBattery:Temp Max: 30 | Temp Min: 27
2024-09-21 21:32:19.084212500 INFO:SerialBattery:Heater 16 Status: False
2024-09-21 21:32:19.084792500 INFO:SerialBattery:Heater 15 Status: False
2024-09-21 21:32:19.085320500 INFO:SerialBattery:===== DVCC State =====
2024-09-21 21:32:19.085869500 INFO:SerialBattery:DVCC Charger Mode: None
2024-09-21 21:32:19.086468500 INFO:SerialBattery:DVCC Charge Voltage: Nonev
2024-09-21 21:32:19.087059500 INFO:SerialBattery:Charge Current: None | Discharge Current: None
2024-09-21 21:32:19.087631500 INFO:SerialBattery:Charge Limit: None | Discharge Limit: None
2024-09-21 21:32:19.088187500 INFO:SerialBattery:===== BMS Data =====
2024-09-21 21:32:19.088856500 INFO:SerialBattery:Voltage: 27.201v | Current: -1.76A
2024-09-21 21:32:19.089522500 INFO:SerialBattery:Capacity Left: 398 of 400.0 AH
2024-09-21 21:32:19.090134500 INFO:SerialBattery:SoC: 100.0%
2024-09-21 21:32:19.090767500 INFO:SerialBattery:SoH: 100.0% | Cycle Count: 1
2024-09-21 21:32:19.091902500 INFO:SerialBattery:Balancing State: UNKNOWN
2024-09-21 21:32:19.092597500 INFO:SerialBattery:===== Warning/Alarms =====
2024-09-21 21:32:19.093224500 INFO:SerialBattery:  No Warnings - 0000
2024-09-21 21:32:19.093818500 INFO:SerialBattery:  No Protection Events - 0000
2024-09-21 21:32:19.094409500 INFO:SerialBattery:  No Errors - 0000
2024-09-21 21:32:19.094925500 INFO:SerialBattery:===== Pack Details =====
2024-09-21 21:32:19.095512500 INFO:SerialBattery:  === BMS ID-16 ===
2024-09-21 21:32:19.096295500 INFO:SerialBattery:  State: Discharging
2024-09-21 21:32:19.096962500 INFO:SerialBattery:  Pack Balancing: Finished
2024-09-21 21:32:19.097695500 INFO:SerialBattery:  Pack Voltage: 27.183v | Pack Current: -0.93a
2024-09-21 21:32:19.098464500 INFO:SerialBattery:    = Cell Stats =
2024-09-21 21:32:19.099115500 INFO:SerialBattery:  Cell 1 Voltage: 3.4
2024-09-21 21:32:19.100048500 INFO:SerialBattery:  Cell 2 Voltage: 3.396
2024-09-21 21:32:19.100774500 INFO:SerialBattery:  Cell 3 Voltage: 3.385
2024-09-21 21:32:19.101585500 INFO:SerialBattery:  Cell 4 Voltage: 3.401
2024-09-21 21:32:19.102291500 INFO:SerialBattery:  Cell 5 Voltage: 3.399
2024-09-21 21:32:19.103027500 INFO:SerialBattery:  Cell 6 Voltage: 3.401
2024-09-21 21:32:19.104305500 INFO:SerialBattery:  Cell 7 Voltage: 3.4
2024-09-21 21:32:19.104689500 INFO:SerialBattery:  Cell 8 Voltage: 3.401
2024-09-21 21:32:19.105484500 INFO:SerialBattery:  Cell Max/Min/Diff: (3.401/3.385/0.016)v
2024-09-21 21:32:19.106080500 INFO:SerialBattery:  === BMS ID-15 ===
2024-09-21 21:32:19.106628500 INFO:SerialBattery:  State: Discharging
2024-09-21 21:32:19.107156500 INFO:SerialBattery:  Pack Balancing: Finished
2024-09-21 21:32:19.107830500 INFO:SerialBattery:  Pack Voltage: 27.22v | Pack Current: -0.83a
2024-09-21 21:32:19.108354500 INFO:SerialBattery:    = Cell Stats =
2024-09-21 21:32:19.109046500 INFO:SerialBattery:  Cell 1 Voltage: 3.389
2024-09-21 21:32:19.109632500 INFO:SerialBattery:  Cell 2 Voltage: 3.406
2024-09-21 21:32:19.110257500 INFO:SerialBattery:  Cell 3 Voltage: 3.406
2024-09-21 21:32:19.110863500 INFO:SerialBattery:  Cell 4 Voltage: 3.402
2024-09-21 21:32:19.111667500 INFO:SerialBattery:  Cell 5 Voltage: 3.405
2024-09-21 21:32:19.112357500 INFO:SerialBattery:  Cell 6 Voltage: 3.406
2024-09-21 21:32:19.112997500 INFO:SerialBattery:  Cell 7 Voltage: 3.406
2024-09-21 21:32:19.113607500 INFO:SerialBattery:  Cell 8 Voltage: 3.4
2024-09-21 21:32:19.114286500 INFO:SerialBattery:  Cell Max/Min/Diff: (3.406/3.389/0.017)v
2024-09-21 21:32:19.140046500 INFO:SerialBattery:publish config values = False
2024-09-21 21:32:19.142154500 INFO:SerialBattery:Battery EG4 LL connected to dbus from /dev/ttyUSB0
2024-09-21 21:32:19.142823500 INFO:SerialBattery:========== Settings ==========
2024-09-21 21:32:19.143532500 INFO:SerialBattery:> Connection voltage: 27.201 V | Current: -1.76 A | SoC: 100.0%
2024-09-21 21:32:19.144110500 INFO:SerialBattery:> Cell count: 8 | Cells populated: 8
2024-09-21 21:32:19.144667500 INFO:SerialBattery:> LINEAR LIMITATION ENABLE: True
2024-09-21 21:32:19.145305500 INFO:SerialBattery:> MIN CELL VOLTAGE: 2.900 V | MAX CELL VOLTAGE: 3.500 V| FLOAT CELL VOLTAGE: 3.375 V
2024-09-21 21:32:19.145949500 INFO:SerialBattery:> MAX BATTERY CHARGE CURRENT: 100.0 A | MAX BATTERY DISCHARGE CURRENT: 100.0 A
2024-09-21 21:32:19.146534500 INFO:SerialBattery:> CVCM:     True
2024-09-21 21:32:19.147161500 INFO:SerialBattery:> CCCM CV:  True  | DCCM CV:  True
2024-09-21 21:32:19.147768500 INFO:SerialBattery:> CCCM T:   True  | DCCM T:   True
2024-09-21 21:32:19.148339500 INFO:SerialBattery:> CCCM SOC: False | DCCM SOC: False
2024-09-21 21:32:19.148899500 INFO:SerialBattery:> CHARGE FET: None | DISCHARGE FET: None | BALANCE FET: None
2024-09-21 21:32:19.149437500 INFO:SerialBattery:Serial Number/Unique Identifier: 4S12400190500001
2024-09-21 21:32:24.504567500 INFO:SerialBattery:===== HW Info =====
**2024-09-21 21:32:24.515132500 Traceback (most recent call last):
2024-09-21 21:32:24.515141500   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 780, in publish_battery
2024-09-21 21:32:24.515144500     result = self.battery.refresh_data()
2024-09-21 21:32:24.515147500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 152, in refresh_data
2024-09-21 21:32:24.515150500     result = self.read_battery_bank()
2024-09-21 21:32:24.515152500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 531, in read_battery_bank
2024-09-21 21:32:24.515286500     self.status_logger(self.battery_stats)
2024-09-21 21:32:24.515289500   File "/opt/victronenergy/dbus-serialbattery/bms/eg4_ll.py", line 307, in status_logger
2024-09-21 21:32:24.515293500     logger.info(f"Battery Make/Model: {self.battery_stats[self.batteryMasterId]['hw_make']}")
2024-09-21 21:32:24.515296500 KeyError: 'hw_make'**
2024-09-21 21:32:41.313527500 INFO:SerialBattery:
2024-09-21 21:32:41.314230500 INFO:SerialBattery:Starting dbus-serialbattery

### batteryPackId = [ 16, 15 ]
statuslogger = True
Comment Out Lines 307, 308, 311

2024-09-21 21:46:40.876959500 INFO:SerialBattery:Starting dbus-serialbattery
2024-09-21 21:46:40.878978500 INFO:SerialBattery:Venus OS v3.42
2024-09-21 21:46:40.879662500 INFO:SerialBattery:dbus-serialbattery v1.3.20240705
2024-09-21 21:46:56.892249500 INFO:SerialBattery:-- Testing BMS: 1 of 3 rounds
2024-09-21 21:46:56.893321500 INFO:SerialBattery:Testing Daly at address "\x40"
2024-09-21 21:46:57.026458500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:46:57.027482500 INFO:SerialBattery:Testing Daly at address "\x80"
2024-09-21 21:46:57.160361500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:46:57.161088500 INFO:SerialBattery:Testing Ecs
2024-09-21 21:46:57.219667500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:46:57.220375500 INFO:SerialBattery:Testing EG4_Lifepower
2024-09-21 21:46:57.502080500 ERROR:SerialBattery:>>> ERROR: No reply - returning
2024-09-21 21:46:57.503342500 INFO:SerialBattery:Testing EG4_LL at address "\x7c"
2024-09-21 21:46:58.539285500 ERROR:SerialBattery:No Reply - BMS ID:15 Command-Hardware
2024-09-21 21:46:58.540565500 INFO:SerialBattery:Connected to BMS ID's: {16: True}
2024-09-21 21:46:58.862885500 INFO:SerialBattery:Connection established to EG4_LL
2024-09-21 21:47:02.025360500 INFO:SerialBattery:Found existing battery with DeviceInstance = 4
2024-09-21 21:47:02.276690500 INFO:SerialBattery:DeviceInstance = 4
2024-09-21 21:47:02.278549500 INFO:SerialBattery:PID file created successfully: /var/tmp/dbus-serialbattery_4.pid
2024-09-21 21:47:02.279341500 INFO:SerialBattery:Used DeviceInstances = ['4', '3', '2', '1']
2024-09-21 21:47:02.280131500 INFO:SerialBattery:com.victronenergy.battery.ttyUSB0
2024-09-21 21:47:03.075413500 INFO:SerialBattery:===== HW Info =====
2024-09-21 21:47:03.076101500 INFO:SerialBattery:Commented out Lines 307, 308, 311
2024-09-21 21:47:03.076677500 INFO:SerialBattery:Commented out Lines 307, 308, 311
2024-09-21 21:47:03.077227500 INFO:SerialBattery:===== Temp =====
2024-09-21 21:47:03.077874500 INFO:SerialBattery:Temp 1: 30c | Temp 2: 28c | Temp Mos: 28c
2024-09-21 21:47:03.078484500 INFO:SerialBattery:Temp Max: 30 | Temp Min: 28
2024-09-21 21:47:03.079105500 INFO:SerialBattery:Heater 16 Status: False
2024-09-21 21:47:03.079683500 INFO:SerialBattery:Heater 15 Status: False
2024-09-21 21:47:03.080209500 INFO:SerialBattery:===== DVCC State =====
2024-09-21 21:47:03.080756500 INFO:SerialBattery:DVCC Charger Mode: None
2024-09-21 21:47:03.081995500 INFO:SerialBattery:DVCC Charge Voltage: Nonev
2024-09-21 21:47:03.082724500 INFO:SerialBattery:Charge Current: None | Discharge Current: None
2024-09-21 21:47:03.083342500 INFO:SerialBattery:Charge Limit: None | Discharge Limit: None
2024-09-21 21:47:03.083951500 INFO:SerialBattery:===== BMS Data =====
2024-09-21 21:47:03.084592500 INFO:SerialBattery:Voltage: 27.921v | Current: 0.0A
2024-09-21 21:47:03.085212500 INFO:SerialBattery:Capacity Left: 400 of 400.0 AH
2024-09-21 21:47:03.085793500 INFO:SerialBattery:SoC: 100.0%
2024-09-21 21:47:03.086380500 INFO:SerialBattery:SoH: 100.0% | Cycle Count: 1
2024-09-21 21:47:03.086926500 INFO:SerialBattery:Balancing State: UNKNOWN
2024-09-21 21:47:03.087473500 INFO:SerialBattery:===== Warning/Alarms =====
2024-09-21 21:47:03.088080500 INFO:SerialBattery:  No Warnings - 0000
2024-09-21 21:47:03.088687500 INFO:SerialBattery:  Protection: 1000 - Low Capacity
2024-09-21 21:47:03.089267500 INFO:SerialBattery:  No Errors - 0000
2024-09-21 21:47:03.089792500 INFO:SerialBattery:===== Pack Details =====
2024-09-21 21:47:03.090445500 INFO:SerialBattery:  === BMS ID-16 ===
2024-09-21 21:47:03.090949500 INFO:SerialBattery:  State: Standby
2024-09-21 21:47:03.091641500 INFO:SerialBattery:  Pack Balancing: Finished
2024-09-21 21:47:03.092365500 INFO:SerialBattery:  Pack Voltage: 27.92v | Pack Current: 0.0a
2024-09-21 21:47:03.092920500 INFO:SerialBattery:    = Cell Stats =
2024-09-21 21:47:03.093539500 INFO:SerialBattery:  Cell 1 Voltage: 3.494
2024-09-21 21:47:03.094137500 INFO:SerialBattery:  Cell 2 Voltage: 3.494
2024-09-21 21:47:03.094779500 INFO:SerialBattery:  Cell 3 Voltage: 3.457
2024-09-21 21:47:03.095428500 INFO:SerialBattery:  Cell 4 Voltage: 3.496
2024-09-21 21:47:03.096097500 INFO:SerialBattery:  Cell 5 Voltage: 3.493
2024-09-21 21:47:03.096739500 INFO:SerialBattery:  Cell 6 Voltage: 3.496
2024-09-21 21:47:03.097391500 INFO:SerialBattery:  Cell 7 Voltage: 3.494
2024-09-21 21:47:03.098064500 INFO:SerialBattery:  Cell 8 Voltage: 3.496
2024-09-21 21:47:03.098766500 INFO:SerialBattery:  Cell Max/Min/Diff: (3.496/3.457/0.039)v
2024-09-21 21:47:03.099385500 INFO:SerialBattery:  === BMS ID-15 ===
2024-09-21 21:47:03.099978500 INFO:SerialBattery:  State: Standby
2024-09-21 21:47:03.100543500 INFO:SerialBattery:  Pack Balancing: Finished
2024-09-21 21:47:03.101355500 INFO:SerialBattery:  Pack Voltage: 27.923v | Pack Current: 0.0a
2024-09-21 21:47:03.101992500 INFO:SerialBattery:    = Cell Stats =
2024-09-21 21:47:03.102665500 INFO:SerialBattery:  Cell 1 Voltage: 3.456
2024-09-21 21:47:03.103290500 INFO:SerialBattery:  Cell 2 Voltage: 3.496
2024-09-21 21:47:03.104274500 INFO:SerialBattery:  Cell 3 Voltage: 3.496
2024-09-21 21:47:03.104916500 INFO:SerialBattery:  Cell 4 Voltage: 3.494
2024-09-21 21:47:03.105600500 INFO:SerialBattery:  Cell 5 Voltage: 3.495
2024-09-21 21:47:03.106247500 INFO:SerialBattery:  Cell 6 Voltage: 3.496
2024-09-21 21:47:03.106922500 INFO:SerialBattery:  Cell 7 Voltage: 3.496
2024-09-21 21:47:03.107664500 INFO:SerialBattery:  Cell 8 Voltage: 3.494
2024-09-21 21:47:03.108410500 INFO:SerialBattery:  Cell Max/Min/Diff: (3.496/3.456/0.04)v


So it appears that both 16 and 15 are being picked up. But I'm not sure what i should be looking for on the Cerbo Remote Console.  
![Screenshot 2024-09-21 175844](https://github.com/user-attachments/assets/3a2e53d1-6bd0-4c1b-acde-ea1e8e7e0571)
![Screenshot 2024-09-21 175901](https://github.com/user-attachments/assets/1540a601-bfac-4ade-bb7e-646c79a810ec)
![Screenshot 2024-09-21 175923](https://github.com/user-attachments/assets/0a061168-2cae-4304-9231-9a371379be45)

I do see that the capacity has updated. 

Should all cells be listed ? Thanks again for the work and all the help.

Cheers

Nathan

**Comment by tuxntoast** [comment]
Few Things: 
	- Looks like you have the V2 units. I know there are differences in the BMS that was used in these units, not sure what / how different they really are. I am testing with the V1 units (no LCD screen on the front) 
	- Looks like the DIP switches are a little different in numbering too. You setting the driver to look for BMS ID of 16 and 15 is working, but looking at your DIP switches they are the same as mine, but that resolved to be ID of 16 and 1. 
	- On the Cerbo, you will not see the cell voltage for both BMS. Right now only the Cell voltage for the master is being displayed. Cerbo is only got support for a single battery to be reported, If I push all the cell voltage, the Cerbo doubled the voltage of the battery and through off DVCC when it is telling other devices that charge voltage they should be using. 
		- I am working to find a way around this, but currently this is how it stands. And the only way to see what current cell voltage is on another other BMS in the chain other then the master is to Flip status logger = True. 

	- Re Logs - batteryPackId = [ 16, 15 ] statuslogger = True Comment Out Lines 307, 308, 311
		- Both batteries are being polled, and reporting data correctly. The error “hw_make” in the prior logs is an item I have addressed already, but may not have pushed that to this repo yet. Will re-push code to address this error soon.

**Comment by tuxntoast** [comment]
Looking over your logs again, I see you had a warning that was caught by the driver for Low Capacity, however the driver also found your cell to be at 3.4v (not LOW!). 

Are you getting Protection Warnings / Alarms? If so, there could be differences in the error codes between v1 & v2, and logic would need to be added to catch the model number, and use that to define the table of what the hex codes are to what they actually represent to the BMS it is polling.

**Comment by Richygecko** [comment]
- Thanks for clarifying that! The dip switches are the same, I had it set to ID=1 in the picture not ID=15 (UUUU).  
- I was indeed getting a low capacity error. Reset the BMS and it went away.
- I'm interested in knowing why you are rolling up the battery values instead of using one of the other aggregator solutions? Any benefit?  I ask because in my application i will need data from each battery, that will then be output to ve.can/nmea2k for other purposes. I've made the unique_identifier unique by adding the following and put each bank into a seperate RS485 to USB adapter. it's working at the moment.  Just curious. 

```
def unique_identifier(self):
        return "4S12400190500001_" + re.sub('[^0-9]','', self.port)
```

Cheers

Nathan

**Comment by tuxntoast** [comment]
Serial management was the biggest; These BMS’s are kind of temperamental on serial. Call it too many times and it will lock up and have to be power cycled, not reply, or start mixing responses. 
		- By managing them all in one driver, I cam make sure that all the calls are timed correctly, and all polling is being graceful on the battery array. 
I also wanted to avoid having more than one communication cable from the Cerbo to the battery array. 

	- My system is in a RV, and both batteries are wired in parallel. Both units are right on top of each other. Really didn’t see a need to have to keep them as separate devices in Cerbo other then getting Cell Voltage per Pack. Code right now is polling each of the Battery Packs, and then either avg / adding / or finding the Min / Max of all the units (where relevant) then reporting that to the Cerbo. So the Cerbo is aware of any concerning values, and just make the monitoring simpler. 

If I really want to track current on a per unit, I can just during on the Statuslogger Switch. Error codes are pulled from all devices, and the unique ID are passed to the Cerbo for reporting. 

Can I ask why you see a need to still use an aggregator to monitor your array of 2? (Other than per cell voltage of all the packs). 

Re unique_id, yes. The driver is actually fetching the “serial” from the BMS, but I had some timing issues of when the device is first connecting, and needed to just hard code the string. I also found out the serial number is not so much unique to the device. The output of StatusLogger, is the serial number that is actually on each of your devices. I had appended the BMS ID to that number to address this non-unique serial number issue. This is bug that I need to revisit now that some other parts are working correctly

**Comment by Richygecko** [comment]
Ok thanks, I understand about serial being temperamental!  Would have been nice to have the updated EG4ll's or 48's with the software update for CAN connection. ( Still trying to figure out what the green rj45 Battery COM connection even does? )

I'm also installed in a RV / Campervan. I'm working on a IoT system that uses NMEA2k / CAN communication from Cerbo and I wanted to be able to see all the cell data for each battery. But not sure if it's worth it. DVCC working correctly would be much more important. And it doesn't appear to transmit the cell data anyway.

I noticed in BMS Tools the ability to change the BMS Serial # but wasn't sure if that was going to affect something else down the road.

**Comment by tuxntoast** [comment]
You talking about the port to the left of CAM? I think that was a special port for their inverters to talk to the array. 

Do you have a real cerbo, or are you running with an arduino? I had that same thoughts when I was mapping out my build, but VRM gives me just about all that I would really want from the data side. With some different tanks sensors, I was able to cover the rest of the systems (fuel / water / waste).  If you have a cerbo, you get free access to VRM!

I looked into NEMA2K, but was looking like a lot of extra hardware / software, for little in return. Makes sense on a large boat, where you want to tap into the Nav / System Control. 

yes, DVCC is used if there is more then one charging source, and keeping them all working together to charge your batteries correctly.

**Comment by Richygecko** [comment]
Hi,

I'm back playing with the batteries and cerbo  and just wondering what the difference between this egy-ll version and the full dbus-serialbattery package is? Any reason to use one or the other?

Cheers
Nathan

**Comment by tuxntoast** [comment]
I have an open PR to get this merged into the main release of dbusserial battery, but right now what is in their release only supports a single battery, the version on this repo allows for a single cable from the cerbo to the master battery, then daisy chaining each other bms that way. The version on this repo also allow for per bms reporting to cerbo. Where the other ia doing a summery any rollup. Meaning you could not get details from all cells. Eventually this will be merged into that main release, if you dont want to wait for that you can just replace this file with was is installed in the cerbo. Sent from my iPhoneOn Apr 18, 2025, at 8:51 AM, Richygecko ***@***.***> wrote:﻿
  Hi,
I'm back playing with the batteries and cerbo  and just wondering what the difference between this egy-ll version and the full dbus-serialbattery package is? Any reason to use one or the other?
Cheers
Nathan—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>



Richygecko left a comment (tuxntoast/eg4-ll#2)
Hi,
I'm back playing with the batteries and cerbo  and just wondering what the difference between this egy-ll version and the full dbus-serialbattery package is? Any reason to use one or the other?
Cheers
Nathan
